editors/code/src/bootstrap.ts

```diff
@@ -52,24 +52,39 @@ async function getServer(
     }
     if (packageJson.releaseTag === null) return "rust-analyzer";
 
-    if (vscode.workspace.workspaceFolders?.length === 1) {
-        // otherwise check if there is a toolchain override for the current vscode workspace
-        // and if the toolchain of this override has a rust-analyzer component
-        // if so, use the rust-analyzer component
-        const toolchainTomlExists = await fileExists(
-            vscode.Uri.joinPath(vscode.workspace.workspaceFolders[0]!.uri, "rust-toolchain.toml"),
-        );
-        if (toolchainTomlExists) {
-            const res = spawnSync("rustup", ["which", "rust-analyzer"], {
-                encoding: "utf8",
-                env: { ...process.env },
-                cwd: vscode.workspace.workspaceFolders[0]!.uri.fsPath,
-            });
-            if (!res.error && res.status === 0) {
-                return res.stdout.trim();
+    let toolchainServerPath = undefined;
+    if (vscode.workspace.workspaceFolders) {
+        for (const workspaceFolder of vscode.workspace.workspaceFolders) {
+            // otherwise check if there is a toolchain override for the current vscode workspace
+            // and if the toolchain of this override has a rust-analyzer component
+            // if so, use the rust-analyzer component
+            const toolchainTomlExists = await fileExists(
+                vscode.Uri.joinPath(workspaceFolder.uri, "rust-toolchain.toml"),
+            );
+            if (toolchainTomlExists) {
+                const res = spawnSync("rustup", ["which", "rust-analyzer"], {
+                    encoding: "utf8",
+                    env: { ...process.env },
+                    cwd: workspaceFolder.uri.fsPath,
+                });
+                if (!res.error && res.status === 0) {
+                    toolchainServerPath = earliestToolchainPath(toolchainServerPath, res.stdout.trim(), function (path) {
+                        const res = spawnSync(path, ["version"], {
```

Comment:
Wouldn't it make more sense to query the rustup version for the workspace here?

Reply:
This is a lambda where the `path` comes from within evaluating the toolchain path. It is invoked conditionally depending on the type of path to the RA it is i.e. if the RA isn't pointing to a nightly or stable one, then the path is deemed to be an explicity version of RA.

Also, this is a lambda for two reasons:

* other `spawnSync` calls are made within this `getServer` function; and
* so that I can write unit tests easily for `earliestToolchainPath` and `orderFromPath`.

Reply:
I'm confused, `rustup which rust-analyzer` should only yield a rust toolchain version (if installed in the toolchain) or error no?

Reply:
It appears to yield paths with stable, nightly and [some-version] in them. At least in my workspace. I’ll create some unit tests with sample paths to illustrate as I move forward.

Reply:
Or may be I don’t understand what you’re asking… if we are talking about the lambda still, that path can be a path to a nightly, stable or versioned RA.

Reply:
Oh I see, we can return a manual (non-toolchain component) rust-analyzer install from this. Okay then this handling does make sense

Reply:
Or actually, what is being invoked here, `path/to/rust-analyzer version`? That wouldn't make sense as that is not valid command for r-a. The flag is `--version`. Can you add a comment above that explaining roughly whats being executed for visibility? Also for the regex part a comment showing an example input and what we disassemble that to via the regexes(regices?) would be helpful for readability

Reply:
Absolutely, I’ll get on to some unit tests tomorrow which should make things a bit more obvious  Also, the test infra is there already (npm test).

I just wanted to make sure I was on the right track etc. Thanks for the initial review.